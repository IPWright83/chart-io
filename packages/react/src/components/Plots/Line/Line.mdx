import { Meta, Story, Canvas } from "@storybook/addon-docs/blocks";

# Line Component

The `<Line>` component allows us to define a line plot on our chart.

<Canvas>
    <Story id="xycharts-line--basic" />
</Canvas>

### Props

Also includes all props from [XY Plot Components API](XYPlot.md)

| Prop           | Type          | Default            | Note                                                                                                         |
| -------------- | ------------- | ------------------ | ------------------------------------------------------------------------------------------------------------ |
| `color`        | `string`      | Derived from Theme | The colour of the data.                                                                                      |
| `layer`\*      | `HTMLElement` |                    | This is an _internal_ field and should not need setting manually. It is populated `withSVG` or `withCanvas`. |
| `x`\*          | `string`      | null               | The key of the field that contains the value along the x-axis.                                               |
| `y`\*          | `string`      | null               | The key of the field that contains the value along the y-axis.                                               |
| `onMouseOver`  | `Function`    | `() => {}`         | Event handler to call when the mouse enter a DataPoint.                                                      |
| `onMouseOut`   | `Function`    | `() => {}`         | Event handler to call when the mouse exits a DataPoint.                                                      |
| `onMouseClick` | `Function`    | `() => {}`         | Event handler to call when the mouse clicks a DataPoint.                                                     |

## Visual Styling

Most of the configuration for a `<Line>` is similar to other plots. Currently the only styling property avaliable is the colour option.

<Canvas>
    <Story id="xycharts-scatter--color" />
</Canvas>

## Multiple Series

There are two ways to go about defining multiple line plots. The simplest way is to use the `<Lines>` (plural) component. The `<Lines>` component allows us to define multiple series at once and takes care of automaticallying layering and assigning colors.

The long hand way is to define each `<Line>`, and configure with a colour manually. The only advantage to this approach is if you wish to control other settings (such as events) differently between the plots.

The results of either approach, render in exactly the same way:

<Canvas>
    <Story id="xycharts-line--multiple-lines" />
</Canvas>

#### Scale

It is however important to remember whichever approach you use to insert any additional keys that you are using into your `<YAxis>` or `<YScale>` component.

`<YAxis fields={["Total Profit", "Total Revenue", "Total Cost"]} />`
