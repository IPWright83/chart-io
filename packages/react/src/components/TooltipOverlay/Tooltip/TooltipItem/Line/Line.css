.chart-it.indicator-line {
    width: 10px;
    height: 3px;
    margin-top: 7px;
    margin-left: 5px;
    margin-right: 5px;
}
