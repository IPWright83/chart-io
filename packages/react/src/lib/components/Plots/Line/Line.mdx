import { Meta, Story, Canvas } from "@storybook/addon-docs/blocks";

# Line Plots

There are two different Line based components, these include:

-   `<Line>` - allows us to define a line plot on our chart
-   `<Lines>` - allows us to define multiple line series

> **Note**: To ensure interactivity this component should be used with the `<EventRecorder>` component which picks up events on the background and pipes them to the plot.

## `<Line>` Component

The `<Line>` component should be used when you want to plot a single line plot on a chart.

<Canvas>
    <Story id="xycharts-line--basic" />
</Canvas>

### Props

| Prop          | Type      | Default            | Note                                                                                                         |
| ------------- | --------- | ------------------ | ------------------------------------------------------------------------------------------------------------ |
| `color`       | `string`  | Derived from Theme | Override the colour of the series                                                                            |
| **`x`\***     | `string`  | null               | The key of the field that contains the value along the x-axis.                                               |
| **`y`\***     | `string`  | null               | The key of the field that contains the value along the y-axis.                                               |
| `interactive` | `boolean` | true               | Whether the plot should be interactive. Setting this to false will disable tooltips & markers from this plot |

### Visual Styling

Most of the configuration for a `<Line>` is similar to other plots. Currently the only styling property avaliable is the colour option.

<Canvas>
    <Story id="xycharts-scatter--color" />
</Canvas>

## `<Lines> Component`

The `<Lines>` component allows us to define multiple line series. When plotting multiple line series they do not stack, but always behave by overlaying on top of one another.

### Props

| Prop       | Type       | Default            | Note                                                           |
| ---------- | ---------- | ------------------ | -------------------------------------------------------------- |
| `colors`   | `string[]` | Derived from Theme | Override the colours for the series                            |
| **`x`\***  | `string`   | null               | The key of the field that contains the value along the x-axis. |
| **`ys`\*** | `string[]` | null               | The keys of the fields that contains values along the y-axis.  |

### Overlapping Lines

The simplest way is to use the `<Lines>` (plural) component. The `<Lines>` component allows us to define multiple series at once and takes care of automaticallying layering and assigning colors. It is however possible to do this manually if required by defining multiple `<Line>` components, configuring with a colour manually.

The results of either approach, render in exactly the same way:

<Canvas>
    <Story id="xycharts-area--multiple-areas" />
</Canvas>
