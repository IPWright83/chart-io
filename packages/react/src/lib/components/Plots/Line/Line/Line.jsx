import PropTypes from "prop-types";
import React from "react";

import { withCanvas, withSVG, withXYPlot } from "../../../../hoc";

import { CanvasLine } from "./CanvasLine";
import { SVGLine } from "./SVGLine";

const WrappedCanvasLine = withCanvas(withXYPlot(CanvasLine), "plot line");
const WrappedSVGLine = withSVG(withXYPlot(SVGLine), "plot line");

/**
 * Represents a Line plot
 * @param  {Object} props       The set of React properties
 * @return {ReactDOMComponent}  The Line plot component
 */
const Line = ({ useCanvas, ...props }) => {
    if (useCanvas) {
        return <WrappedCanvasLine {...props} />;
    }

    return <WrappedSVGLine {...props} />;
};

Line.propTypes = {
    ...SVGLine.propTypes,

    /**
     * Should this plot use an HTML Canvas instead of SVG?
     * @type {[type]}
     */
    useCanvas: PropTypes.bool,
};

Line.requiresVirtualCanvas = false;

export { Line };
