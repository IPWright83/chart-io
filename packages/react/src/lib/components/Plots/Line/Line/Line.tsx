import type { IPlotProps } from "@chart-io/types";
import React from "react";

import { withCanvas, withSVG, withXYPlot } from "../../../../hoc";

import { CanvasLine } from "./CanvasLine";
import { SVGLine } from "./SVGLine";

export interface ILineProps extends Omit<IPlotProps, "layer" | "canvas"> {
    /**
     * Should Canvas be used instead of SVG?
     */
    useCanvas?: boolean;
}

const WrappedCanvasLine = withCanvas(withXYPlot<ILineProps>(CanvasLine), "plot line");
const WrappedSVGLine = withSVG(withXYPlot<ILineProps>(SVGLine), "plot line");

/**
 * Represents a Line plot
 * @param  props       The set of React properties
 * @return             The Line plot component
 */
export function Line({ useCanvas = false, ...props }: ILineProps) {
    if (useCanvas) {
        return <WrappedCanvasLine {...props} />;
    }

    return <WrappedSVGLine {...props} />;
}

Line.requiresVirtualCanvas = false;
Line.isPlot = true;
Line.brush = {
    horizontal: true,
    vertical: false,
};
Line.zoom = {
    horizontal: true,
    vertical: false,
};
