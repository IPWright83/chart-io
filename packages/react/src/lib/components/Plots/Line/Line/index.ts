export * from "./Line";
export * from "./CanvasLine";
export * from "./SVGLine";
