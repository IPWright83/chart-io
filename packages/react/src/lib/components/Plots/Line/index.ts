export * from "./Line";
export * from "./Lines";
