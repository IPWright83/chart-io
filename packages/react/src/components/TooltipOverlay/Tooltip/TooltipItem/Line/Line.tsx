import React from "react";
import PropTypes from "prop-types";

import type { IColor } from "../../../../../types";

import "./Line.css";

export interface ILineProps {
    /**
     * The fill color of the Line
     * @type {String}
     */
    fill: IColor;
}

const Line = ({ fill }: ILineProps) => {
    const style = {
        background: fill,
    };

    return <div className="chart-it indicator-line" style={style} />;
};

export { Line };
