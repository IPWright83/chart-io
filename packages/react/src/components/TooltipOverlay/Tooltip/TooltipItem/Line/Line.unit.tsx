import React from "react";
import { render } from "@testing-library/react";

import { Line } from "./Line";

describe("Line", () => {
    it("renders correctly", () => {
        const { asFragment } = render(<Line fill="#4682B4" />);

        expect(asFragment()).toMatchSnapshot();
    });
});
