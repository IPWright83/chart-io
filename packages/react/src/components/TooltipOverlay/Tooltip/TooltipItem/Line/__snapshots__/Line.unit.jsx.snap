// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Line renders correctly 1`] = `
<DocumentFragment>
  <div
    class="chart-it indicator-line"
    style="background: steelblue;"
  />
</DocumentFragment>
`;
